tools/flight_recorder/fr_trace.py

```diff
@@ -235,42 +297,79 @@ def __repr__(self) -> str:
             )
         return f"{self.type}(input_sizes={self.input_sizes}, {self.state})"
 
-    def match(self, other) -> bool:  # type: ignore[no-untyped-def]
+    def match(self, other: "Op") -> MatchState:
         # TODO: I think this can validly not match,
         # e.g. if one PG was used for p2p ops between only some of the peers?
         # if self.seq_id != other.seq_id:
         # return False
 
         if self.type == "send":
-            return bool(
-                other.type == "recv"
-                and self.src == other.src
-                and self.dst == other.dst
-                and self.input_sizes == other.output_sizes
+            # TODO: We need more states for p2p ops.
+            return (
+                MatchState.FULLY_MATCHED
+                if (
+                    other.type == "recv"
+                    and self.src == other.src
+                    and self.dst == other.dst
+                    and self.input_sizes == other.output_sizes
+                )
+                else MatchState.SIZE_OR_SYNTAX_MISMATCH
             )
         elif self.type == "recv":
-            return bool(
-                other.type == "send"
-                and self.src == other.src
-                and self.dst == other.dst
-                and self.output_sizes == other.input_sizes
+            return (
+                MatchState.FULLY_MATCHED
+                if (
+                    other.type == "send"
+                    and self.src == other.src
+                    and self.dst == other.dst
+                    and self.output_sizes == other.input_sizes
+                )
+                else MatchState.SIZE_OR_SYNTAX_MISMATCH
             )
         elif self.type in COLLECTIVES:
-            return bool(
-                self.type == other.type and self.input_sizes == other.input_sizes
-            )
-            # TODO(whc) - output sizes dont have to match for e.g. gather, not sure if they ever have to match?
-            # and self.output_sizes == other.output_sizes)
+            if self.type != other.type:
+                return MatchState.COLLECTIVE_TYPE_MISMATCH
+            if self.type in ["alltoall", "alltoall_base"]:
+                return MatchState.UNDECIDED
+            if self.type != "scatter" and self.input_sizes != other.input_sizes:
+                return MatchState.SIZE_OR_SYNTAX_MISMATCH
```

Comment:
nit: feels like we could capture even more information in the MatchState object.

I wonder if we should actually construct an instance of MatchState and return that.  The instance could contain a 'mismatch string' or some lists of items A and items B where later on, someone analyzing the MatchState can see what the mismatch specifically was.

Reply:
Got it, sure. We can do that.